Fix/jspsych 5.0

zenkavi opened the pull request:
updated jspsych to the most recent version (5.0 - https://github.com/jodeleeuw/jsPsych/releases/tag/v5.0) and modified custom plugins to be compatible with the new version. also created a simple md for the changes that need to be to the experiment.js folders within expfactory/expfactory-experiments: https://github.com/zenkavi/expfactory-battery/blob/fix/jspsych-5.0/static/js/jspsych-5.0/jspsych5.0-change-notes.md

vsoch commented:
holy cow this is diabolical! I will take a look at the md file first thing tomorrow so we can update the battery and such before merging the fixed experiments. Great job! :D

zenkavi commented:
Not sure how helpful the md will be to follow the changes made to the repo. They were more haphazard notes on things we will have to do to update expfactory/experiments experiment.js files. Most of the added code is not mine but just installs most recent version of jspsych within static/js/jspsych-5.0. Basically the plugins directory has been replaced. Based on this I updated the customs_plugins to comply with the new format. I also changed the html files that I believe (though am not sure) are being called when running a batter or previewing a task

vsoch commented:
Ok, I see. Let's do this for a game plan:

- Add a top header with a date to the change notes file. This is fantastic documentation, and if there are changes in the future we can use the file to document such changes.
- The jspsych.css file should not be in a css subfolder in jspsych, it should be in the static/css folder
- the tests and examples folder should not be included if we are not using it. It has a lot of larger files (images, etc) that will just take up space and not be used.
- I think the folder should still be called "jspsych" to support the idea that our experiments should be updated to work with the most recent version that we choose. On the one hand, we could decide to support legacy scripts for a package, and then we would want folders with versions. But since we are releasing fresh, I think we should take an approach that asserts to keep experiments updated with the most recent software, and they should break if not updated. What we would need is a way to better test function, which is probably outside the scope of this initial pilot.

If you want to make those changes, what I can do is merge this PR, and could you provide detail about changes to the (old) way of running the experiments? Since we have some of this code in expfactory-python and expfactory-docker, I would want to update those. Then I can regenerate our web interface expfactory.github.io and we can see how many experiments are broken, and go from there :)

What we really need is a robust way to test function of javascript. I know there are testing frameworks but I'm not familiar with any - this seems important but less of a priority than getting it working, period. For now we can do manual testing, but let's keep this in mind. I'll add an issue for it.

@IanEisenberg would be good to get your thoughts on the above!

IanEisenberg commented:
I agree with everything. I'll move the css file go the static css folder.
The way we have it now we contain the jspsych package in one place, but
that doesn't make sense given our template organization.

I agree we should keep the folder 'jspsych'. If in the future we inevitably
upgrade jspsych and experiments are not backwards compatible and we have to
fix a bunch of stuff - yes, we should keep a legacy '5.0' version
somewhere. Let's start with tracking those versions after 5.0.

I'm not sure what the change notes file is.

On Tue, Jan 12, 2016 at 10:02 AM, Vanessa Sochat <notifications@github.com>
wrote:

> Ok, I see. Let's do this for a game plan:
>
>    - Add a top header with a date to the change notes file. This is
>    fantastic documentation, and if there are changes in the future we can use
>    the file to document such changes.
>    - The jspsych.css file should not be in a css subfolder in jspsych, it
>    should be in the static/css folder
>    - the tests and examples folder should not be included if we are not
>    using it. It has a lot of larger files (images, etc) that will just take up
>    space and not be used.
>    - I think the folder should still be called "jspsych" to support the
>    idea that our experiments should be updated to work with the most recent
>    version that we choose. On the one hand, we could decide to support legacy
>    scripts for a package, and then we would want folders with versions. But
>    since we are releasing fresh, I think we should take an approach that
>    asserts to keep experiments updated with the most recent software, and they
>    should break if not updated. What we would need is a way to better test
>    function, which is probably outside the scope of this initial pilot.
>
> If you want to make those changes, what I can do is merge this PR, and
> could you provide detail about changes to the (old) way of running the
> experiments? Since we have some of this code in expfactory-python and
> expfactory-docker, I would want to update those. Then I can regenerate our
> web interface expfactory.github.io and we can see how many experiments
> are broken, and go from there :)
>
> What we really need is a robust way to test function of javascript. I know
> there are testing frameworks but I'm not familiar with any - this seems
> important but less of a priority than getting it working, period. For now
> we can do manual testing, but let's keep this in mind. I'll add an issue
> for it.
>
> @IanEisenberg <https://github.com/IanEisenberg> would be good to get your
> thoughts on the above!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/expfactory/expfactory-battery/pull/11#issuecomment-170995927>
> .
>



-- 
Best,
Ian

vsoch commented:
[This one](https://github.com/zenkavi/expfactory-battery/blob/fix/jspsych-5.0/static/js/jspsych-5.0/jspsych5.0-change-notes.md)

I'm zooming into making the changes to the docker instance this morning, so expect the site to go up and down. Now that we have https working the rest of the stuff should be easy. I forgot to write down the board in Russ office, so check the [issues that I created](https://github.com/expfactory/expfactory-docker/issues) to see if I forgot anything!

Also FYI - the active branch is dev/aws, the master is pretty old at this point!

zenkavi commented:
Agree with almost all the points. I'll make the the changes for the first three bullet points and update the pr.
Def agree w testing more rigorously but know little about it so an issue for now sounds good.
Re folder name for jspsych: i actually think we should have it as 5 to help developers know what version they are on. Though we should strive to keep everything on the most recent version jspsych operates independent of us and it won't always be backwards compatible (as it was this time) and it might not always be feasible to immediately update all experiments. I won't push on this too strongly but in my own hacky development setup right now having the two directories separately named by version has been helpful

vsoch commented:
For the naming, I'm just thinking of what should be the default when we update a version. If we always change the name of the folder, that means enormous amounts of work to go through templates, experiments, etc, and manually change the version name. If we keep the folder as jspsych, then we don't need to do that, and we use tests to find experiments that need updating/fixing. I would say that it would be better to have the current version as jspsych, and then support backwards compatibility (eg, have a folder jspych4.0 that people could change their experiment to if they did not want the update).

vsoch commented:
For the naming, I'm just thinking of what should be the default when we update a version. If we always change the name of the folder, that means enormous amounts of work to go through templates, experiments, etc, and manually change the version name. If we keep the folder as jspsych, then we don't need to do that, and we use tests to find experiments that need updating/fixing. I would say that it would be better to have the current version as jspsych, and then support backwards compatibility (eg, have a folder jspych4.0 that people could change their experiment to if they did not want the update).

zenkavi commented:
As usual more forward thinking than me. Got it now. Will change that as well and push it all in a bit

zenkavi commented:
@vsoch i think i made all the changes you listed. please let me know if i have missed any others

vsoch commented on exp_hub.html:
```diff
@@ -58,7 +58,8 @@
 		jsPsych.init({
-		  experiment_structure: experiments,
+		  //experiment_structure: experiments,
+		  timeline: experiments,
```
We didn't have timeline before - what exactly is this?

IanEisenberg replied:
The timeline is the replacement for another variabel "experiment" or
something like that. In this case it is just a variable rename.

On Tue, Jan 12, 2016 at 1:25 PM, Vanessa Sochat <notifications@github.com>
wrote:

> In exp_hub.html
> <https://github.com/expfactory/expfactory-battery/pull/11#discussion_r49515945>
> :
>
> > @@ -58,7 +58,8 @@
> >
> >  		// start the experiment
> >  		jsPsych.init({
> > -		  experiment_structure: experiments,
> > +		  //experiment_structure: experiments,
> > +		  timeline: experiments,
>
> We didn't have timeline before - what exactly is this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/expfactory/expfactory-battery/pull/11/files#r49515945>
> .
>



-- 
Best,
Ian

vsoch replied:
ok I'll need to update [here as well](https://github.com/expfactory/expfactory-docker/blob/master/expdj/apps/turk/templates/mturk_battery.html#L72)
